src/main/scala/org/apache/spark/sql/execution/datasources/greenplum/GreenplumUtils.scala

```diff
@@ -111,51 +111,57 @@ object GreenplumUtils extends Logging {
     val suffix = "sparkGpTmp"
     val quote = "\""
 
-    val tempTable = s"${schemaPrefix}$quote${rawTblName}_${randomString}_${suffix}$quote"
+    val tempTable = s"$schemaPrefix$quote${rawTblName}_${randomString}_$suffix$quote"
     val strSchema = JdbcUtils.schemaString(df, options.url, options.createTableColumnTypes)
     val createTempTbl = s"CREATE TABLE $tempTable ($strSchema) ${options.createTableOptions}"
 
+    // Stage 1. create a _sparkGpTmp table as a shadow of the target Greenplum table. If this stage
+    // fails, the whole process will abort.
     val conn = JdbcUtils.createConnectionFactory(options)()
-    var transactionSuccessful = false
-    var tempTblCreated = false
-
     try {
       executeStatement(conn, createTempTbl)
-      tempTblCreated = true
-      val accumulator = df.sparkSession.sparkContext.longAccumulator("copySuccess")
-      val partNum = df.rdd.getNumPartitions
+    } finally {
+      closeConnSilent(conn)
+    }
 
-      df.foreachPartition { rows =>
-        copyPartition(rows, options, schema, tempTable, Some(accumulator))
-      }
+    // Stage 2. Spark executors run copy task to Greenplum, and will increase the accumulator if
+    // each task successfully copied.
+    val accumulator = df.sparkSession.sparkContext.longAccumulator("copySuccess")
+    df.foreachPartition { rows =>
+      copyPartition(rows, options, schema, tempTable, Some(accumulator))
+    }
 
+    // Stage 3. if the accumulator value is not equal to the [[Dataframe]] instance's partition
+    // number. The Spark job will fail with a [[PartitionCopyFailureException]].
+    // Otherwise, we will run a rename table ddl statement to rename the tmp table to the final
+    // target table.
+    val partNum = df.rdd.getNumPartitions
+    val conn2 = JdbcUtils.createConnectionFactory(options)()
+    try {
       if (accumulator.value == partNum) {
-        if (JdbcUtils.tableExists(conn, options)) {
-          JdbcUtils.dropTable(conn, options.table)
+        if (JdbcUtils.tableExists(conn2, options)) {
+          JdbcUtils.dropTable(conn2, options.table)
         }
 
         val newTableName = s"${options.table}".split("\\.").last
-        val renameTempTbl = s"ALTER TABLE $tempTable RENAME TO ${newTableName}"
-        executeStatement(conn, renameTempTbl)
-        transactionSuccessful = true
+        val renameTempTbl = s"ALTER TABLE $tempTable RENAME TO $newTableName"
+        executeStatement(conn2, renameTempTbl)
```

Comment:
If exception occurred during `dropTable` and `renameTable`, we also need  drop the temp table.

Reply:
fixed